File: Java/Substring with Concatenation of All Words.java
public class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        if (s == null || s.length() == 0 || 
        words == null || words.length == 0) {
            return new ArrayList<Integer>();
        }
        
        List<Integer> result = new ArrayList<Integer>();
        
        HashMap<String, Integer> map = new HashMap<String, Integer>();
        for (int i=0; i<words.length; i++) {
            if (map.containsKey(words[i])) {
                map.put(words[i], map.get(words[i])+1);
            } else {
                map.put(words[i], 1);
            }
        }
        int len = words[0].length();
        int count = words.length;
        char[] chs = s.toCharArray();
        
        for (int i=0; i+len*count<=s.length(); i++) {
            int p = len + i;
            int subLen = 0;
            HashMap<String, Integer> m = new HashMap<String, Integer>();
            while (subLen < len*count) {
                String subStr = new String(chs, p-len, len);
                if (map.containsKey(subStr)) {
                    int num = map.get(subStr);
                    if (m.containsKey(subStr)) {
                        int mNum = m.get(subStr);
                        if (mNum+1 <= num) {
                            m.put(subStr, mNum+1);
                            p += len;
                            subLen += len;    
                        } else {
                            break;
                        }
                    } else {
                        m.put(subStr, 1);
                        p += len;
                        subLen += len;
                    }
                } else {
                    break;
                }
            }
            if (subLen == len*count) {
                result.add(i);
            }
        }
        
        return result;
    }
}

File: Python/Substring with Concatenation of All Words.py
class Solution(object):
    """
    :type s: str
    :type words: List[str]
    :rtype: List[int]
    """
    def findSubstring(self, s, words):
        if s == None or len(s) == 0 or \
            words == None or len(words) == 0:
                return []
                
        result = []
        toFind = {}
        for str in words:
            if str in toFind:
                toFind[str] = toFind[str] + 1
            else:
                toFind[str] = 1
                
        wordLen = len(words[0])
        count   = len(words)
        
        for i in range(0, len(s)-wordLen*count+1, 1):
            found = {}
            for j in range(0, count, 1):
                k = i + j*wordLen
                subStr = s[k:k+wordLen]
                if subStr not in toFind:
                    break
                if subStr not in found:
                    found[subStr] = 1
                else:
                    found[subStr] = found[subStr] + 1
                if found[subStr] > toFind[subStr]:
                    break
            else:
                result.append(i)
        
        return result            